params.go

```diff
@@ -0,0 +1,157 @@
+package slog
+
+import (
+	"context"
+	"sync"
+	"testing"
+)
+
+// WithParams returns a copy of the parent context containing the given log parameters.
+// Any log events generated using the returned context will include these parameters
+// as metadata.
+//
+// For example:
+//
+//	 ctx := slog.WithParams(ctx, map[string]string{
+//	   "foo_id": fooID,
+//	   "bar_id": barID,
+//	 })
+//
+//	slog.Info(ctx, "Linking foo to bar")  // includes foo_id and bar_id parameters
+//
+// If the parent context already contains parameters set by a previous call to WithParams,
+// the new parameters will be merged with the existing set, with newer values taking
+// precedence over older ones.
+//
+// It is not safe to modify the supplied map after passing it to WithParams.
```

Comment:
ooi why not copy the map "on the way in" as well?

Reply:
Only that I find it much harder to imagine cases where the caller wants to mutate the input map _after_ passing it to `slog.WithParams`, whereas mutating the output map feels relatively likely to me:

```go
errParams := slog.Params(ctx)
errParams["internal_error"] = err.Error()
return terrors.Unauthorized("foo", "", errParams)
```

... though perhaps I just lack imagination.

For reference, cloning a 3-parameter map takes roughly 100ns on my machine.